### PR TITLE
dependencies: compute and check release dates via GitHub API.

### PR DESCRIPTION
**Pull request by htuch**
The use of last_updated was ambiguous (is it when an Envoy contributor
creates a PR, merges a commit, or when the dependency is released?).
We really are after the release date as a measure of how stale the
dependency is.

This patch introduces a tool, tools/dependency/release_dates.py, that
uses the GitHub API to compute release date. If a mismatch is detected, an
error is raised.

This patch also introduces a dependency validation CI job that gathers existing
scripts and the release_dates.py script into a single job.

Co-authored-by: Michael Payne <michael@sooper.org>
Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
CC @PiotrSikora @envoyproxy/dependency-shepherds

**Comment by htuch** [comment]
@PiotrSikora @danzh2010 @dio can you update on whether the `last_updated` dates on the release bundles live on GCS and belong to the com_googlesource_* dependencies that you introduced or maintain reflect "last updated in Envoy" or "release date"? thanks.

**Comment by PiotrSikora** [comment]
@htuch V8's release date should be 2020-08-17 (for 8.5.210.20), BoringSSL's release date is correct.

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -682,7 +682,7 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         urls = ["https://storage.googleapis.com/quiche-envoy-integration/googleurl_{version}.tar.gz"],
         use_category = ["dataplane_ext"],
         extensions = [],
-        last_updated = "2020-08-05",
+        release_date = "2020-08-05",
```
Comment (dio) [inline bazel/repository_locations.bzl:685]: The release date: 2020-07-30.

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -669,7 +669,7 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         urls = ["https://storage.googleapis.com/quiche-envoy-integration/{version}.tar.gz"],
         use_category = ["dataplane_ext"],
         extensions = ["envoy.transport_sockets.quic"],
-        last_updated = "2020-09-18",
+        release_date = "2020-09-18",
```
Comment (danzh2010) [inline bazel/repository_locations.bzl:672]: This date is the date I updated the tar ball. The release date of this commit should be 2020-09-17

**Review by moderation — DISMISSED** [review]
LGTM

**Comment by htuch** [comment]
@lizan this is ready for review; it's stuck on not having GH_ACCESS_TOKEN set, since we need to populate Azure with an access token for "public repo read-only" access to the API.

**Comment by htuch** [comment]
This PR now passes the new CI "precheck dependencies" :)

File: ci/check_repository_locations.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+set -e
+
+function no_change {
+  echo "No change to **/repository_locations.bzl"
+  exit 0
+}
+
+(./tools/git/modified_since_last_github_commit.sh . bzl | grep repository_locations) || no_change
+
+./tools/dependency/release_dates.sh  ./bazel/repository_locations.bzl
```
Comment (lizan) [inline ci/check_repository_locations.sh:12]: nit: extra space

File: ci/run_envoy_docker.sh [hunk 4]
```diff
@@ -86,6 +86,7 @@ docker run --rm \
        -e SYSTEM_PULLREQUEST_TARGETBRANCH \
        -e SYSTEM_PULLREQUEST_PULLREQUESTNUMBER \
        -e GCS_ARTIFACT_BUCKET \
+       -e GH_ACCESS_TOKEN \
```
Comment (lizan) [inline ci/run_envoy_docker.sh:89]: name this `GITHUB_TOKEN`? That is already a widely used name.

Comment (htuch) [reply ci/run_envoy_docker.sh]: Makes sense. I've also renamed other uses of GH_ACCESS_TOKEN to GITHUB_TOKEN in the repo.

**Review by lizan — APPROVED** [review]

**Comment by htuch** [comment]
MacOS failures looks RBE related, https://dev.azure.com/cncf/envoy/_build/results?buildId=54800&view=logs&j=b4895c6e-031d-5a9e-a507-86b9350b5c70&t=b4ab6edc-d679-5a1d-980f-d72dcbda6eb1. Will force merge.

**Pull request merged**